2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+---
+RFC: RFCxxxx
+Author: Steve Lee
+Status: Draft
+SupercededBy: N/A
+Version: 0.1
+Area: PowerShellGet
+Comments Due: 7/30
+Plan to implement: Yes
+---
+
+# PowerShellGet 3.0 Module
+
+PowerShellGet has enabled an active community to publish and install PowerShell
+Modules.
+However, it has some architectural and design limitations that prevent it from
+moving forward more quickly.
+This RFC proposes some significant changes to address this.
+
+> [!NOTE]
+> There are no intentions to break existing PSGet v2 users so existing
+> users can stay on PSGet v2 and that will continue to work.
+
+## Motivation
+
+    As a PowerShell user,
+    I can discover how to install missing cmdlets,
+    so that I can be more productive.
+
+    As a PowerShellGet contributor,
+    I can more easily contribute code,
+    so that PowerShellGet evolves more quickly.
+
+## User Experience
+
+```powershell
+PS> Get-Satisfaction
+Get-Satisfaction : The term 'Get-Satisfaction' is not recognized as the name of a cmdlet, function, script file, or operable program.
+Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
+At line:1 char:1
++ Get-Satisfaction
++ ~~~
++ CategoryInfo          : ObjectNotFound: (Get-Satisfaction:String) [], CommandNotFoundException
++ FullyQualifiedErrorId : CommandNotFoundException
+
+Suggestion: You can obtain `Get-Satisfaction` by running `Install-PSResource HappyDays`.
+
+PS> Install-PSResource HappyDays
+```
+
+## Specification
+
+### Rewrite of PowerShellGet
+
+PowerShellGet is currently written as PowerShell script with a dependency on PackageManagement.
+Proposal is to write PSResource in C# to reduce complexity and make easier to maintain.
```

Comment:
The crux of the problem with PowerShellGet is the dependency on Package Management and the mixing of C# and PowerShell based codebases between them and the sheer level of extreme defensive programming in PowerShellGet

I'm not sure that I agree that rewriting PowerShellGet in C# to reduce complexity and make it easier to maintain is technically correct, I mean it may be from a PowerShell Team perspective due to the fact that historically the PowerShell team have been more comfortable writing C# than PowerShell, but from a community perspective and longer term maintainability I think this may be a 50/50 as they both have their merits in differing ways.  However that's my view as someone more comfortable writing PowerShell than C# - others may have differing views on this

Reply:
The complexity is due to mixing PowerShell and C# as well as the general complexity of PkgMgmt itself which resulted in complex PowerShell script code in PSGet.  You're right that from a pure code perspective, rewriting it in PowerShell script sans PkgMgmt would be similar to C# for someone well versed in PowerShell script.  As you note, there is a practical matter for the engineers I have available to implement this.

Reply:
Hey, there's always https://github.com/vexx32/PSKoans 😉

Reply:
Ultimately using C# will probably mean we can fine tune some things better, specifically with regard to performance and if any threading is required